fix(terminal): interrupt/got_int hangs terminal

ibhagwan opened the pull request:
`terminal_execute` should always clear `got_int` or `safe_vgetc` will keep returning `Ctrl_C` in an endless loop.

Minimal reproduction of the issue:
> will cause neovim to hang with 100% CPU prior to this patch
```vim
:autocmd TermEnter * call timer_start(500, {-> interrupt()})
:new | term
:startinsert
```

Closes #20726

ibhagwan committed:
fix: terminal hang on interrupt

ref: https://github.com/neovim/neovim/issues/20726

Upon `terminal_enter`, `mapped_ctrl_c` is set in order to avoid `CTRL-C`
interrupts (which is proxied to the terminal process instead), `os_inchar`
will then test `mapped_ctrl_c` against `State` and set `ctrl_c_interrupts=false`
which prevents `process_ctrl_c` from setting `got_int=true` in a terminal
state.

However, if `got_int` is set outside of `process_ctrl_c`, e.g. via
`interrupt()`, this will hang the neovim process as `terminal_execute` will
enter an endless loop as `got_int` will never be cleared causing `safe_vgetc`
to always return `Ctrl_C`.

A minimal example reproducing this bug:
```vim
:autocmd TermEnter * call timer_start(500, {-> interrupt()})
:terminal
:startinsert
```

To fix, we make sure `got_int` is cleared inside `terminal_execute` when
it detects `Ctrl_C`.

Closes #20726

justinmk commented:
please includes the relevant parts of your analysis, in the commit message. and also the comment from https://github.com/neovim/neovim/issues/20726#issuecomment-2290574907 is helpful: 

> The are some other places that can set got_int. For example, the hang can happen when calling interrupt() in Terminal mode.
> So got_int needs to be cleared more often

justinmk commented on src/nvim/terminal.c:
```diff
@@ -726,6 +726,9 @@ static int terminal_execute(VimState *state, int key)
 
   default:
+    if (key == Ctrl_C) {
+      got_int = false;
```
how is anyone going to know the purpose of this in the future? git spelunking which leads to... a gigantic github discussion... is costly. We can easily compress the conclusion into a minimal comment.

```suggestion
      // terminal_enter unconditionally sets `mapped_ctrl_c` to avoid got_int. 8eeda7169aa4
      // But got_int may be set elsewhere, e.g. by interrupt(). So ensure that it is cleared.
      got_int = false;
```

justinmk replied:
After this change, do we even need the `mapped_ctrl_c |= ...` line?

ibhagwan replied:
> After this change, do we even need the `mapped_ctrl_c |=` ... line?

For the purpose of `got_int` in terminals, it woild be safe to  assume it is no longer required.

That said, I’m not that acquianted with the code base to guarantee `mapped_ctrl_c` doesn’t have any other purpose, I can do some more digging and see if it’s used anywhere else but since we won’t be able to remove the condition that sets `ctrl_c_interrupts` (as it’s required for buffer ctrl-c mappings) the benefit of removing `mapped_ctrl_c` is not that impactful.

> how is anyone going to know the purpose of this in the future? git spelunking which leads to... a gigantic github discussion... is costly. We can easily compress the conclusion into a minimal comment.

Fair enough, this would be  my first commit to neovim, I am not yet familiar with comment preferences and best practices in this repo.

justinmk commented:
> Minimal reproduction of the issue:

can you add a test? `test/functional/terminal/buffer_spec.lua` looks like a good place

ibhagwan commented:
> > Minimal reproduction of the issue:
> 
> can you add a test? `test/functional/terminal/buffer_spec.lua` looks like a good place

Sure, I’ll have to do some of the testing codebase exploration on how to test that the process isn’t frozen, if you can point me towards an existing similar test would be very helpful and save me some time :)

ibhagwan committed:
ci: added temrminal interrupt test

ref: https://github.com/neovim/neovim/pull/30056

ibhagwan commented:
> Maybe you can just use `feed()` and then `screen:expect()` to check if the process is responding to input.

I wasn't able to run the tests locally yet so I thought of a different way to test stolen from the `ModeChanged` tests, using the autocmd that sends a delayed `interrupt` if the process is frozen we won't be able to return from terminal (insert) mode to normal mode.

https://github.com/neovim/neovim/pull/30056/commits/7e41baf6ec57ee2daff7371b8bddb15ad551e0cf

justinmk commented:
confirmed that the test case fails correctly, though it hangs (so `screen:expect()` would be ideal, since it handles hangs better)

ibhagwan commented:
> confirmed that the test case fails correctly, though it hangs (so `screen:expect()` would be ideal, since it handles hangs better)

Them @justinmk!


> if there are no further comments, ping me in a few days to merge this

Will do, have a great weekend.

zeertzjq committed:
test: simplify test for terminal interrupt

- Remove the timer delay.
- Use nvim_get_mode as it's processed after pending input.

ibhagwan commented:
Tysm @zeertzjq for simplifying the CI!

@justinmk, would be a great start of the week when this gets merged and I can close the only lingering issue in fzf-lua :)

ibhagwan commented:
Ty for all the help @zeertzjq @justinmk walking me through this 🙏
